### PR TITLE
ARROW-10328: [C++] Vendor fast_float number parsing library

### PR DESCRIPTION
**Pull request by pitrou**
This library is 2x to 3x faster for parsing strings to binary floating-point numbers.

### CODE REVIEW

**Comment by pitrou** [comment]
Before:
* Parsing raw values:
```
FloatParsing<FloatType>          39763 ns        39758 ns        52350 items_per_second=25.152M/s
FloatParsing<DoubleType>         29812 ns        29808 ns        68765 items_per_second=33.5483M/s
```
* CSV converter performance:
```
FloatConversion                 199539 ns       199508 ns         3498 items_per_second=40.0986M/s
```
* Reading a CSV file of floating-point numbers (single-threaded):
```
276 MB (10000000 rows) in 1.543 s. => 179 MB/s.
276 MB (10000000 rows) in 1.481 s. => 186 MB/s.
276 MB (10000000 rows) in 1.492 s. => 185 MB/s.
276 MB (10000000 rows) in 1.469 s. => 188 MB/s.
276 MB (10000000 rows) in 1.471 s. => 188 MB/s.
```

**Comment by pitrou** [comment]
After:
* Parsing raw values:
```
FloatParsing<FloatType>          11936 ns        11934 ns       184367 items_per_second=83.7942M/s
FloatParsing<DoubleType>         10989 ns        10988 ns       194016 items_per_second=91.0101M/s
```
* CSV converter performance:
```
FloatConversion                 108180 ns       108166 ns         6442 items_per_second=73.9605M/s
```
* Reading a CSV file of floating-point numbers (single-threaded):
```
276 MB (10000000 rows) in 0.928 s. => 298 MB/s.
276 MB (10000000 rows) in 0.940 s. => 294 MB/s.
276 MB (10000000 rows) in 0.918 s. => 301 MB/s.
276 MB (10000000 rows) in 0.896 s. => 309 MB/s.
276 MB (10000000 rows) in 0.900 s. => 307 MB/s.
```

**Comment by wesm** [comment]
Do we need to unvendor double-conversion?

**Comment by pitrou** [comment]
No, we still use it for the other way round (float to string).

File: LICENSE.txt [hunk 1]
```diff
@@ -2223,3 +2223,11 @@ exception of some code pulled in from other repositories (such as
...
 
 (*) https://creativecommons.org/publicdomain/zero/1.0/legalcode
+
+--------------------------------------------------------------------------------
+
+The files in cpp/src/arrow/vendored/fast_float/ contain code from
+
+https://github.com/lemire/fast_float
+
+which is made available under the Apache License 2.0.
```
Comment (pitrou) [inline LICENSE.txt:2233]: @lemire Is the above blurb acceptable to you, or should we mention a more detailed attribution? Don't hesitate to suggest a wording.

**Review by bkietz — APPROVED** [review]
Nice speedup, I'll wait to merge until attribution has been resolved

File: LICENSE.txt (hunk 1 above)
Comment (lemire) [reply LICENSE.txt:2233]: @pitrou Yes, yes. This is 100% fine.

Comment (pitrou) [reply LICENSE.txt:2233]: Thank you!

Comment (lemire) [reply LICENSE.txt:2233]: @pitrou For future reference, if I post code publicly on GitHub, you can grab it. I am never going to make a fuss.

**Pull request closed without merging**